### PR TITLE
Fix fallout from rust-lang/rust PR 60861

### PR DESCRIPTION
**Pull request by Centril**
Fixes incoming breakage for unlanded https://github.com/rust-lang/rust/pull/60861.

Tests are passing locally; the Rust PR now needs to land first.

@Manishearth also says we'll want to split out to a `collapsible_if_let` once we have let-chains working in Rust nightly or something.

### CODE REVIEW

**Comment by Centril** [comment]
Should be ready to go, I think.

**Comment by lzutao** [comment]
Your build has failed.

**Comment by Centril** [comment]
Let's see if we do better this time.

**Comment by lzutao** [comment]
Formatting error, but it is a little weird (might need to improve on rustfmt side):
```diff
Diff in /Users/travis/build/rust-lang/rust-clippy/clippy_lints/src/needless_continue.rs at line 210:
     F: FnMut(&ast::Block, Option<&ast::Label>),
 {
     if let ast::ExprKind::While(_, loop_block, label)
-        | ast::ExprKind::ForLoop(_, _, loop_block, label)
-        | ast::ExprKind::Loop(loop_block, label)
-        = &expr.node
+    | ast::ExprKind::ForLoop(_, _, loop_block, label)
+    | ast::ExprKind::Loop(loop_block, label) = &expr.node
     {
         func(loop_block, label.as_ref());
     }
```

**Comment by Centril** [comment]
Ugh... I tested this locally with cargo test. :(

File: clippy_lints/src/needless_continue.rs [hunk 1]
```diff
@@ -210,9 +210,8 @@ where
     F: FnMut(&ast::Block, Option<&ast::Label>),
 {
     if let ast::ExprKind::While(_, loop_block, label)
-        | ast::ExprKind::ForLoop(_, _, loop_block, label)
-        | ast::ExprKind::Loop(loop_block, label)
-        = &expr.node
+    | ast::ExprKind::ForLoop(_, _, loop_block, label)
```
Comment (Centril) [inline clippy_lints/src/needless_continue.rs]: This is truly hideous... cc @topecongiro

File: clippy_lints/src/formatting.rs [hunk 2]
```diff
@@ -245,12 +245,10 @@ fn is_block(expr: &ast::Expr) -> bool {
     }
 }
 
-/// Match `if` or `if let` expressions and return the `then` and `else` block.
```
Comment (Manishearth) [inline clippy_lints/src/formatting.rs]: The comment should stay the same

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: we probably should just stop using this function and match, really

Comment (Centril) [reply clippy_lints/src/formatting.rs]: Why? `if let` expressions are not a separate syntactic construct anymore.

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: They are from the point of view of Rustaceans, at least for now.

The implementation is not the same thing as the pedagogical model; which is still the same for now.

File: clippy_lints/src/needless_continue.rs (hunk 1 above)
Comment (topecongiro) [reply clippy_lints/src/needless_continue.rs]: Created https://github.com/rust-lang/rustfmt/issues/3648.

File: clippy_lints/src/formatting.rs (hunk 2 above)
Comment (Centril) [reply clippy_lints/src/formatting.rs]: But this is the implementation (in Clippy), not the pedagogical model (in Clippy docs).

(As for the pedagogical mode, it needs to change with the language.)

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: What I'm saying is that this would not be clear to clippy contributors.

This is in the implementation of _clippy_, not the compiler.

> (As for the pedagogical model, it needs to change with the language.)

But the language hasn't gone there yet. The feature is logically the chaining of if and if let, and that doesn't necessarily mean that we've gotten rid of the distinction. `let` isn't a full fledged expression in this model, there's still a level of distinction.

The bottom line here is that the function exists for a very particular purpose and changing the comment this way makes this purpose unclear.

Please either get rid of the function (it's not really necessary anymore) or fix the comment.

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: Until we're in a situation where pedagogically we teach if and if let as the same kind of thing, _we are not there yet_, and comments in code should reflect the current situation. Not some future ideal. The point of comments is to help those working on the code.

Comment (Centril) [reply clippy_lints/src/formatting.rs]: > Please either get rid of the function (it's not really necessary anymore) or fix the comment.

I've changed back the comment since it is your code-base.

That said, I'm not convinced by what you are saying. I very well understand the point of comments, but I think the change in the comment reflects reality, not some future ideal. Namely, if we keep in mind that AST and HIR are two separate languages (that is, "Rust" is a tower of languages, not a single one), including in terms of semver, then `let` is a truly fully fledged expression in the model of AST.

Moreover, clippy contributors touching this code work on the AST of the compiler, not some other structure, and in AST, `if let` does not exist.

Anyways... let's agree to disagree here.

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: > Moreover, clippy contributors touching this code work on the AST of the compiler, not some other structure

Sure, but the AST/HIR frequently doesn't match the way we want to look at things, that's _why_ we have desugar/unsugar functions in the first place! We talk about things in terms of the actual Rust syntax (i.e. the pedagogical model), not the AST/HIR -- we frequently shim over cases where the AST/HIR doesn't match the pedagogical model so that the code makes sense to someone unfamiliar with the precise corners of the Rust AST.

Clippy uses the AST model because it has to, not because it wants to. We've got efforts to move _away_ from the AST/HIR models, even: https://github.com/rust-lang/rust-clippy/pull/3875/files

Comment (Centril) [reply clippy_lints/src/formatting.rs]: > We talk about things in terms of the actual Rust syntax (i.e. the pedagogical model),

But that's *my* point. 😂 

The *actual* Rust *concrete syntax* supports `let` in arbitrary expression positions on nightly. It's not just a matter of AST representation but what `parser.rs` can produce. Now if you would say that we want to talk about stable Rust, I'd say that would be a good point.

In the specific context of `unsugar_if`, I believe it is also entirely irrelevant whether the condition is a `let` or something else.

Comment (Manishearth) [reply clippy_lints/src/formatting.rs]: > The actual Rust concrete syntax supports let in arbitrary expression positions on nightly

No, by syntax I mean the concepts we use to teach and talk about the syntax, not the actual grammar.

Besides, let in arbitrary positions can parse but is forbidden, it's an implementation detail that the parser accepts it. The RFCd feature only allows let expressions within ifs, and the final pedagogical model for that may not be to consider let to be an expression.

> In the specific context of unsugar_if, I believe it is also entirely irrelevant whether the condition is a let or something else.

Sure, but the reason that function exists is to distinguish itself from checking for _just_ normal `if`s.

(Which is why I said we should remove it)
